agent/agent.go

```diff
@@ -1949,7 +1941,53 @@ func (a *Agent) AddService(service *structs.NodeService, chkTypes []*structs.Che
 		if chkType.Status != "" {
 			check.Status = chkType.Status
 		}
-		if err := a.AddCheck(check, chkType, persist, token, source); err != nil {
+
+		checks = append(checks, check)
+	}
+
+	err := a.State.AddServiceWithChecks(service, checks, token)
+	if err != nil {
+		// we failed, cleanup checks and registrations
+		a.State.RemoveService(service.ID)
+		for _, check := range checks {
+			a.State.RemoveCheck(check.CheckID)
+		}
+		return err
+	}
+
+	for i := range checks {
+		if err := a.addCheck(checks[i], chkTypes[i], service, persist, token, source); err != nil {
+			// we failed, cleanup checks and registrations
+			a.State.RemoveService(service.ID)
+			for _, check := range checks {
+				a.cancelCheckMonitors(check.CheckID)
+				a.State.RemoveCheck(check.CheckID)
+			}
+			return err
+		}
+
+		if persist {
+			if err := a.persistCheck(checks[i], chkTypes[i]); err != nil {
+				a.State.RemoveService(service.ID)
+				for _, check := range checks {
```

Comment:
You can probably factorize the code in a single defer in case of issue.
The defer code has to be clever:

if at t0 we insert:
serviceX(check0)

if at t1, we insert
serviceX(check0, invalidCheck1)

We do not want to revert serviceX and check0, only invalidCheck1